File: go/middleware/guardrails/step_enforcer.go
package guardrails

import "errors"

var ErrStepNotAllowed = errors.New("step not allowed: prerequisite steps must be completed first")

type StepEnforcer struct {
	stepDefinitions  map[string][]string
	completedSteps   map[string]map[string]bool
	allowedTerminals map[string]map[string]bool
}

func NewStepEnforcer() *StepEnforcer {
	return &StepEnforcer{
		stepDefinitions:  make(map[string][]string),
		completedSteps:   make(map[string]map[string]bool),
		allowedTerminals: make(map[string]map[string]bool),
	}
}

func (se *StepEnforcer) AddStep(tool string, prerequisites []string) {
	se.stepDefinitions[tool] = prerequisites
}

func (se *StepEnforcer) AddTerminal(tool string, allowed map[string]bool) {
	se.allowedTerminals[tool] = allowed
}

func (se *StepEnforcer) MarkStepComplete(sessionID, step string) {
	if se.completedSteps[sessionID] == nil {
		se.completedSteps[sessionID] = make(map[string]bool)
	}
	se.completedSteps[sessionID][step] = true
}

func (se *StepEnforcer) ResetSession(sessionID string) {
	delete(se.completedSteps, sessionID)
}

func (se *StepEnforcer) CanExecute(sessionID, tool string) (bool, []string) {
	prereqs, ok := se.stepDefinitions[tool]
	if !ok {
		return true, nil
	}

	completed := se.completedSteps[sessionID]
	if completed == nil {
		completed = make(map[string]bool)
	}

	var missing []string
	for _, prereq := range prereqs {
		if !completed[prereq] {
			missing = append(missing, prereq)
		}
	}

	return len(missing) == 0, missing
}

func (se *StepEnforcer) ValidateExecution(sessionID, tool string) error {
	allowed, missing := se.CanExecute(sessionID, tool)
	if allowed {
		return nil
	}
	return &StepNotAllowedError{
		Tool:         tool,
		MissingSteps: missing,
	}
}

func (se *StepEnforcer) IsTerminalAllowed(sessionID, terminalTool string) bool {
	allowed, _ := se.CanExecute(sessionID, terminalTool)
	return allowed
}

func (se *StepEnforcer) GetAllowedTerminals(sessionID string) []string {
	var result []string
	for tool := range se.stepDefinitions {
		if allowed, _ := se.CanExecute(sessionID, tool); allowed {
			result = append(result, tool)
		}
	}
	return result
}

type StepNotAllowedError struct {
	Tool         string
	MissingSteps []string
}

func (e *StepNotAllowedError) Error() string {
	return "step not allowed"
}

func (e *StepNotAllowedError) Is(target error) bool {
	return target == ErrStepNotAllowed
}

func (e *StepNotAllowedError) Missing() []string {
	return e.MissingSteps
}

File: typescript/src/middleware/guardrails/step_enforcer.ts
export class StepNotAllowedError extends Error {
  tool: string;
  missingSteps: string[];

  constructor(tool: string, missingSteps: string[]) {
    super(`step not allowed: missing ${missingSteps.join(", ")}`);
    this.name = "StepNotAllowedError";
    this.tool = tool;
    this.missingSteps = missingSteps;
  }
}

export class StepEnforcer {
  private stepDefinitions: Map<string, string[]>;
  private completedSteps: Map<string, Map<string, boolean>>;
  private allowedTerminals: Map<string, Map<string, boolean>>;

  constructor() {
    this.stepDefinitions = new Map();
    this.completedSteps = new Map();
    this.allowedTerminals = new Map();
  }

  addStep(tool: string, prerequisites?: string[]): void {
    this.stepDefinitions.set(tool, prerequisites || []);
  }

  addTerminal(tool: string, allowed?: Map<string, boolean>): void {
    this.allowedTerminals.set(tool, allowed || new Map());
  }

  markStepComplete(sessionId: string, step: string): void {
    if (!this.completedSteps.has(sessionId)) {
      this.completedSteps.set(sessionId, new Map());
    }
    this.completedSteps.get(sessionId)!.set(step, true);
  }

  resetSession(sessionId: string): void {
    this.completedSteps.delete(sessionId);
  }

  canExecute(sessionId: string, tool: string): [boolean, string[]] {
    const prereqs = this.stepDefinitions.get(tool);
    if (prereqs === undefined) {
      return [true, []];
    }

    const completed = this.completedSteps.get(sessionId) || new Map();
    const missing = prereqs.filter((p) => completed.get(p) !== true);

    return [missing.length === 0, missing];
  }

  validateExecution(sessionId: string, tool: string): void {
    const [allowed, missing] = this.canExecute(sessionId, tool);
    if (allowed) {
      return;
    }
    throw new StepNotAllowedError(tool, missing);
  }

  isTerminalAllowed(sessionId: string, terminalTool: string): boolean {
    const [allowed] = this.canExecute(sessionId, terminalTool);
    return allowed;
  }

  getAllowedTerminals(sessionId: string): string[] {
    const result: string[] = [];
    for (const tool of this.stepDefinitions.keys()) {
      const [allowed] = this.canExecute(sessionId, tool);
      if (allowed) {
        result.push(tool);
      }
    }
    return result;
  }
}